src/integrations/watchers/indexable-ancestor-watcher.php

```diff
@@ -88,14 +110,105 @@ public function reset_children( $indexable, $indexable_before ) {
+
+		\array_walk( $child_indexables, [ $this, 'refresh_permalink' ] );
 
-			$child->permalink = $this->indexable_helper->get_permalink_for_indexable( $child );
-			$child->save();
+		if ( $indexable->object_type === 'term' ) {
+			$child_indexables_for_term = $this->get_children_for_term( $indexable->object_id, $child_indexables );
+
+			\array_walk( $child_indexables_for_term, [ $this, 'refresh_permalink' ] );
 		}
 
 		return true;
 	}
+
+	/**
+	 * Finds all child indexables for the given term.
+	 *
+	 * @param int         $term_id          Term to fetch the indexable for.
+	 * @param Indexable[] $child_indexables The already known child indexables.
+	 *
+	 * @return array The list of additional indexables for a given term.
+	 */
+	public function get_children_for_term( $term_id, array $child_indexables ) {
+		// Finds object_ids (posts) for the term.
+		$object_ids = $this->get_object_ids_for_term( $term_id );
+
+		// Removes the objects that are already present in the children.
+		foreach ( $child_indexables as $child_indexable ) {
+			if ( $child_indexable->object_type !== 'post' ) {
+				continue;
+			}
+
+			$search = \array_search( $child_indexable->object_id, $object_ids, true );
+			if ( ! $search ) {
+				continue;
+			}
+
+			unset( $object_ids[ $search ] );
+		}
+
+		// Finds the indexables for the fetched object_ids.
+		$indexables_by_term = $this->indexable_repository->find_by_multiple_ids_and_type( $object_ids, 'post' , false );
+
+		// Finds the indexables for the posts that attached to the term.
+		$additional_indexable_ids = $this->indexable_hierarchy_repository->find_children_by_ancestor_ids( $object_ids );
+
+		// Makes sure we only have indexable ids that we haven't fetched before.
+		foreach ( $indexables_by_term as $indexable_by_term ) {
+			$search = \array_search( $indexable_by_term->id, $additional_indexable_ids, true );
+			if ( ! $search ) {
+				continue;
+			}
+
+			unset( $additional_indexable_ids[ $search ] );
+		}
+
+		// Finds the additional indexables.
+		$additional_indexables = $this->indexable_repository->find_by_ids( $additional_indexable_ids );
+
+		// Lets merge all fetched indexables.
+		return \array_merge( $indexables_by_term, $additional_indexables );
+	}
+
+	/**
+	 * Refreshes the permalink for the given indexable.
+	 *
+	 * @param Indexable $indexable The indexable to refresh the permalink for.
+	 */
+	protected function refresh_permalink( $indexable ) {
+		$this->indexable_hierarchy_builder->build( $indexable );
+
+		$indexable->permalink = $this->indexable_helper->get_permalink_for_indexable( $indexable );
+		$indexable->save();
+	}
+
+	/**
+	 * Retrieves the object ids for a term based on the term-post relation ship.
+	 *
+	 * @param int $term_id The term to get the object ids for.
+	 *
+	 * @return array List with object ids for the term.
+	 */
+	protected function get_object_ids_for_term( $term_id ) {
+		$term_taxonomy_ids = $this->wpdb->get_col(
```

Comment:
Done. We didn't check whether the child indexables "are terms" because the indexable hierarchy table only marks other terms as children of terms, right? Therefore, in the case of a term, `$child_indexable_ids = $this->indexable_hierarchy_repository->find_children( $indexable );` should only contain children that are terms.

Reply:
Unfortunately not, a post without a post parent that does have a category has that category as it's parent.

In the end the hierarchy is based off the breadcrumbs, so each 'parent' in the hierarchy is essentially the breadcrumb parent.

However you're already retrieved the full indexable objects so you can simply run an `array_filter` on the `$child_indexables` and check if `object_type` is `term`.